### ISSUE TITLE
primitives: Make `hashes` optional

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3087]
#3039 makes `hashes` an unconditional dependency of `primitives`, before release we should consider using an inner array for the new types created with `hash_newtype` and making `hashes` optional.

**Comment by Kixunil** [comment]
That implies privately copying the `hash_newtype` macro and making it output the same/stable code. While I dislike copying a big chunk of code I think it might be needed and even the safe thing to do so we don't accidentally break its output.

**Comment by apoelstra** [comment]
I think the `hash_newtype` macro ought to be renamed anyway to `byte_array_wrapper` or something, and that `hashes` isn't the right home for it.

**Comment by Kixunil** [comment]
I was thinking about it too and I think it's true, especially now that the `Hash` trait is going to be removed from the type. The reverse option is kinda weird but isn't too bad. I'd put it into `internals` for now.

**Comment by tcharding** [comment]
This issue is now discussing two things

1. Should `hashes` be an optional dependency of `primitives`
2. Is the `hash_newtype` macro misnamed and should it be moved to another place

Re (2) we also have `bitcoin::internal_macros::impl_array_newtype` but the API produced differs from `hash_newtype` (even just the byte-array bits eg `AsRefMut`) 


> That implies privately copying the hash_newtype macro and making it output the same/stable code

This is not entirely true, the issue is meant to describe the idea of putting a plain old array inside the new hashtypes if the feature is not enabled eg `Txid([u8; 32])` instead of `Txid(sha256::Hash)`. That was your idea if I"ve understood it correctly @apoelstra. Are you still keen on the idea?

**Comment by apoelstra** [comment]
I'm still keen on the idea, but given that we failed to remove the `Hash` trait (I don't remember why now..) I'm a little less optimistic that it'll be possible.

**Comment by sanket1729** [comment]
Is the primary motivation for splitting hashes from primitives, compilation time and granularity? Another alternative is to feature flags for each of the hashes (sha2, sha1, sha3, hash160) etc. I think the crypto libraries from rust-crypto do this. primitives can then only the necessary hashes it needs.

**Comment by apoelstra** [comment]
Good idea. I thnk we should do both. A further motivation to make the whole dep optional is that it eliminates a whole crate from what users need to audit, and in principle reduces the attack surface. (Though in practice it won't since `hashes` and `primitives` have the same crates.io owners and live in the same git repo.)

**Comment by tcharding** [comment]
> I thnk we should do both.

What did you mean here by 'both'? Are you referring to 'compilation time' and 'graularity'?

FWIW, assuming compilation time is proportional to line count, the `hashes` crate has ~5800 LOC and we would need at least ~2600 of them in `primitives`. I do not know much about compilation times, is that line reduction worth the effort?

**Comment by Kixunil** [comment]
Hashes use declarative macros, not proc macros. Those are much much much faster.

**Comment by apoelstra** [comment]
True -- as far as compilation time I don't think `hashes` is meaningfully slower for its macro usage, vs a hypothetical version of `hashes` which had all its code laid out explicitly.
